File: core/src/main/java/com/oneandone/typedrest/BlobEndpoint.java
package com.oneandone.typedrest;

import java.io.*;
import java.util.Optional;
import org.apache.http.*;
import org.apache.http.entity.ContentType;

/**
 * REST endpoint that represents a single binary blob that can downloaded and
 * uploaded.
 */
public interface BlobEndpoint extends Endpoint {

    /**
     * Queries the server about capabilities of the endpoint without performing
     * any action.
     *
     * @throws IOException Network communication failed.
     * @throws IllegalArgumentException {@link HttpStatus#SC_BAD_REQUEST}
     * @throws IllegalAccessException {@link HttpStatus#SC_UNAUTHORIZED} or
     * {@link HttpStatus#SC_FORBIDDEN}
     * @throws FileNotFoundException {@link HttpStatus#SC_NOT_FOUND} or
     * {@link HttpStatus#SC_GONE}
     * @throws IllegalStateException {@link HttpStatus#SC_CONFLICT}
     * @throws RuntimeException Other non-success status code.
     */
    void probe()
            throws IOException, IllegalArgumentException, IllegalAccessException, FileNotFoundException, IllegalStateException;

    /**
     * Shows whether the server has indicated that {@link #download()} is
     * currently allowed.
     *
     * Uses cached data from last response.
     *
     * @return An indicator whether the method is allowed. If no request has
     * been sent yet or the server did not specify allowed methods
     * {@link Optional#empty()} is returned.
     */
    Optional<Boolean> isDownloadAllowed();

    /**
     * Downloads the blob's content.
     *
     * @return A stream with the blob's content.
     *
     * @throws IOException Network communication failed.
     * @throws IllegalArgumentException {@link HttpStatus#SC_BAD_REQUEST}
     * @throws IllegalAccessException {@link HttpStatus#SC_UNAUTHORIZED} or
     * {@link HttpStatus#SC_FORBIDDEN}
     * @throws FileNotFoundException {@link HttpStatus#SC_NOT_FOUND} or
     * {@link HttpStatus#SC_GONE}
     * @throws IllegalStateException {@link HttpStatus#SC_CONFLICT}
     * @throws RuntimeException Other non-success status code.
     */
    InputStream download()
            throws IOException, IllegalArgumentException, IllegalAccessException, FileNotFoundException, IllegalStateException;

    /**
     * Shows whether the server has indicated that
     * {@link #upload(java.io.InputStream, org.apache.http.entity.ContentType)}
     * is currently allowed.
     *
     * Uses cached data from last response.
     *
     * @return An indicator whether the method is allowed. If no request has
     * been sent yet or the server did not specify allowed methods
     * {@link Optional#empty()} is returned.
     */
    Optional<Boolean> isUploadAllowed();

    /**
     * Uploads new content for the blob.
     *
     * @param stream A stream with the blob's new content.
     * @param contentType The MIME type of the file to upload.
     *
     * @throws IOException Network communication failed.
     * @throws IllegalArgumentException {@link HttpStatus#SC_BAD_REQUEST}
     * @throws IllegalAccessException {@link HttpStatus#SC_UNAUTHORIZED} or
     * {@link HttpStatus#SC_FORBIDDEN}
     * @throws FileNotFoundException {@link HttpStatus#SC_NOT_FOUND} or
     * {@link HttpStatus#SC_GONE}
     * @throws IllegalStateException {@link HttpStatus#SC_CONFLICT}
     * @throws RuntimeException Other non-success status code.
     */
    void upload(InputStream stream, ContentType contentType)
            throws IOException, IllegalArgumentException, IllegalAccessException, FileNotFoundException;

    /**
     * Uploads a byte array as the blob's content.
     *
     * @param data The byte array holding the data to upload.
     * @param contentType The MIME type of the file to upload.
     *
     * @throws IOException Network communication failed.
     * @throws IllegalArgumentException {@link HttpStatus#SC_BAD_REQUEST}
     * @throws IllegalAccessException {@link HttpStatus#SC_UNAUTHORIZED} or
     * {@link HttpStatus#SC_FORBIDDEN}
     * @throws FileNotFoundException {@link HttpStatus#SC_NOT_FOUND} or
     * {@link HttpStatus#SC_GONE}
     * @throws IllegalStateException {@link HttpStatus#SC_CONFLICT}
     * @throws RuntimeException Other non-success status code.
     */
    void upload(byte[] data, ContentType contentType)
            throws IOException, IllegalArgumentException, IllegalAccessException, FileNotFoundException;

    /**
     * Uploads a local file as the blob's content.
     *
     * @param file The local file to read the data from.
     * @param contentType The MIME type of the file to upload.
     *
     * @throws IOException Network communication failed.
     * @throws IllegalArgumentException {@link HttpStatus#SC_BAD_REQUEST}
     * @throws IllegalAccessException {@link HttpStatus#SC_UNAUTHORIZED} or
     * {@link HttpStatus#SC_FORBIDDEN}
     * @throws FileNotFoundException {@link HttpStatus#SC_NOT_FOUND} or
     * {@link HttpStatus#SC_GONE}
     * @throws IllegalStateException {@link HttpStatus#SC_CONFLICT}
     * @throws RuntimeException Other non-success status code.
     */
    void upload(File file, ContentType contentType)
            throws IOException, IllegalArgumentException, IllegalAccessException, FileNotFoundException, IllegalStateException;

    /**
     * Shows whether the server has indicated that {@link #delete()} is
     * currently allowed.
     *
     * Uses cached data from last response.
     *
     * @return An indicator whether the method is allowed. If no request has
     * been sent yet or the server did not specify allowed methods
     * {@link Optional#empty()} is returned.
     */
    Optional<Boolean> isDeleteAllowed();

    /**
     * Deletes the blob from the server.
     *
     * @throws IOException Network communication failed.
     * @throws IllegalArgumentException {@link HttpStatus#SC_BAD_REQUEST}
     * @throws IllegalAccessException {@link HttpStatus#SC_UNAUTHORIZED} or
     * {@link HttpStatus#SC_FORBIDDEN}
     * @throws FileNotFoundException {@link HttpStatus#SC_NOT_FOUND} or
     * {@link HttpStatus#SC_GONE}
     * @throws IllegalStateException {@link HttpStatus#SC_CONFLICT}
     * @throws RuntimeException Other non-success status code.
     */
    void delete()
            throws IOException, IllegalArgumentException, IllegalAccessException, FileNotFoundException;

}

File: typedrest/src/main/kotlin/net/typedrest/endpoints/raw/BlobEndpoint.kt
package net.typedrest.endpoints.raw

import java.io.InputStream
import net.typedrest.endpoints.*
import net.typedrest.errors.*
import net.typedrest.http.HttpStatusCode
import java.io.*

/**
 * Endpoint for a binary blob that can be downloaded or uploaded.
 */
interface BlobEndpoint : Endpoint {
    /**
     * Queries the server about capabilities of the endpoint without performing any action.
     *
     * @throws AuthenticationException when the server responds with [HttpStatusCode.Unauthorized].
     * @throws AuthorizationException when the server responds with [HttpStatusCode.Forbidden].
     * @throws NotFoundException when the server responds with [HttpStatusCode.NotFound] or [HttpStatusCode.Gone].
     * @throws HttpException for other non-success status codes.
     */
    fun probe()

    /**
     * Indicates whether the server has specified that downloading is currently allowed.
     *
     * Uses cached data from the last response.
     *
     * @return true if the method is allowed, false if the method is not allowed, null if no request has been sent yet or the server did not specify allowed methods.
     */
    val isDownloadAllowed: Boolean?

    /**
     * Downloads the blob's content to an input stream.
     *
     * @return An input stream with the blob's content.
     * @throws AuthenticationException when the server responds with [HttpStatusCode.Unauthorized].
     * @throws AuthorizationException when the server responds with [HttpStatusCode.Forbidden].
     * @throws NotFoundException when the server responds with [HttpStatusCode.NotFound] or [HttpStatusCode.Gone].
     * @throws HttpException for other non-success status codes.
     */
    fun download(): InputStream

    /**
     * Downloads the blob's content to a file.
     *
     * @param path The path of the file to read the upload data from.
     * @throws IOException when the file at the specified path can not be written.
     * @throws AuthenticationException when the server responds with [HttpStatusCode.Unauthorized].
     * @throws AuthorizationException when the server responds with [HttpStatusCode.Forbidden].
     * @throws HttpException for other non-success status codes.
     */
    fun download(path: String) =
        FileOutputStream(path).use { fileStream ->
            download().use { downloadStream ->
                downloadStream.copyTo(fileStream)
            }
        }

    /**
     * Indicates whether the server has specified that uploading is currently allowed.
     *
     * Uses cached data from the last response.
     *
     * @return true if the method is allowed, false if the method is not allowed, null if no request has been sent yet or the server did not specify allowed methods.
     */
    val isUploadAllowed: Boolean?

    /**
     * Uploads data as the blob's content from an input stream.
     *
     * @param stream The input stream to read the upload data from.
     * @param mimeType The MIME type of the data to upload, nullable.
     * @throws BadRequestException when the server responds with [HttpStatusCode.BadRequest].
     * @throws AuthenticationException when the server responds with [HttpStatusCode.Unauthorized].
     * @throws AuthorizationException when the server responds with [HttpStatusCode.Forbidden].
     * @throws HttpException for other non-success status codes.
     */
    fun uploadFrom(stream: InputStream, mimeType: String? = null)

    /**
     * Uploads content as the blob's content from a file.
     *
     * @param path The path of the file to read the upload data from.
     * @param mimeType The MIME type of the data to upload, nullable.
     * @throws IOException when the file at the specified path can not be read.
     * @throws BadRequestException when the server responds with [HttpStatusCode.BadRequest].
     * @throws AuthenticationException when the server responds with [HttpStatusCode.Unauthorized].
     * @throws AuthorizationException when the server responds with [HttpStatusCode.Forbidden].
     * @throws HttpException for other non-success status codes.
     */
    fun uploadFrom(path: String, mimeType: String? = null) =
        FileInputStream(path).use { fileStream ->
            uploadFrom(fileStream, mimeType)
        }

    /**
     * Indicates whether the server has specified that deleting is currently allowed.
     *
     * Uses cached data from the last response.
     *
     * @return true if the method is allowed, false if the method is not allowed, null if no request has been sent yet or the server did not specify allowed methods.
     */
    val isDeleteAllowed: Boolean?

    /**
     * Deletes the blob from the server.
     *
     * @throws BadRequestException when the server responds with [HttpStatusCode.BadRequest].
     * @throws AuthenticationException when the server responds with [HttpStatusCode.Unauthorized].
     * @throws AuthorizationException when the server responds with [HttpStatusCode.Forbidden].
     * @throws NotFoundException when the server responds with [HttpStatusCode.NotFound] or [HttpStatusCode.Gone].
     * @throws HttpException for other non-success status codes.
     */
    fun delete()
}
